src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2532,160 +2653,246 @@ var projectUniqueNamesForBuildIntToUpdate
+                }
+                else
                 {
-                    // Remove the package from all frameworks and dependencies section.
-                    PackageSpecOperations.RemoveDependency(updatedPackageSpec, action.PackageIdentity.Id);
+                    if (!nugetProjectActionsLookup.ContainsKey(buildIntegratedProject.MSBuildProjectPath))
+                    {
+                        throw new ArgumentNullException($"Either should have value in {nameof(nugetProjectActionsLookup)} for {buildIntegratedProject.MSBuildProjectPath} or {nameof(packageIdentity)} & {nameof(primarySources)}");
+                    }
+
+                    nuGetProjectActions = nugetProjectActionsLookup[buildIntegratedProject.MSBuildProjectPath].ToArray();
+
+                    if (!nuGetProjectActions.Any())
+                    {
+                        // Continue to next project if there are no actions for current project.
+                        continue;
+                    }
                 }
-                else if (action.NuGetProjectActionType == NuGetProjectActionType.Install)
+
+                // Find all sources used in the project actions
+                var sources = new HashSet<SourceRepository>(
+                    nuGetProjectActions.Where(action => action.SourceRepository != null)
+                        .Select(action => action.SourceRepository),
+                        new SourceRepositoryComparer());
+
+                allSources.UnionWith(sources);
+                sources.UnionWith(enabledSources);
+                nuGetProjectSourceLookup[buildIntegratedProject.MSBuildProjectPath] = sources;
+
+                // Read the current lock file if it exists
+                LockFile originalLockFile = null;
+                var lockFileFormat = new LockFileFormat();
+
+                var lockFilePath = await buildIntegratedProject.GetAssetsFilePathAsync();
+
+                if (File.Exists(lockFilePath))
                 {
-                    if (updatedPackageSpec.RestoreMetadata.ProjectStyle == ProjectStyle.PackageReference)
+                    originalLockFile = lockFileFormat.Read(lockFilePath);
+                }
+
+                // Get Package Spec as json object
+                var originalPackageSpec = await DependencyGraphRestoreUtility.GetProjectSpec(buildIntegratedProject, dependencyGraphContext);
+                originalNugetPackageSpecLookup[buildIntegratedProject.MSBuildProjectPath] = originalPackageSpec;
+
+                // Create a copy to avoid modifying the original spec which may be shared.
+                var updatedPackageSpec = originalPackageSpec.Clone();
+
+                // If the lock file does not exist, restore before starting the operations
+                if (originalLockFile == null)
+                {
+                    var originalRestoreResult = await DependencyGraphRestoreUtility.PreviewRestoreAsync(
```

Comment:
What would be the downside if this restore is not done?

Reply:
During `ResolvedAction` execution time it uses it to find addedPackages then run powershell scripts on it.
https://github.com/NuGet/NuGet.Client/blob/24c713cd927fe4920ec5ecc6fe98033f84eab727/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L3041-L3043